src/settlement-chain/PayerReportManager.sol

```diff
@@ -0,0 +1,381 @@
+        address[] calldata nodeOperators_,
+        PayerReportSignature[] calldata signatures_
+    ) internal view returns (uint32[] memory validSigningNodeIds_) {
+        bytes32 digest_ = _getPayerReportDigest(
+            originatorNodeId_,
+            attesterNodeId_,
+            startSequenceId_,
+            endSequenceId_,
+            payersMerkleRoot_,
+            nodeOperators_
+        );
+
+        // Need to determine how many of the signatures are valid and provided by canonical node operators.
+        uint8 validSignatureCount_ = 0;
+
+        // This array will help in constructing the `validSigningNodeIds_` array, who's length may be less than
+        // the length of the `signatures_` array if some of the signatures are invalid.
+        bool[] memory isValid_ = new bool[](signatures_.length);
+
+        for (uint256 index_; index_ < signatures_.length; ++index_) {
+            uint32 nodeId_ = signatures_[index_].nodeId;
+
+            // Enforces that the signing node IDs are ordered and unique.
+            if (index_ != 0 && nodeId_ <= signatures_[index_ - 1].nodeId) revert UnorderedNodeIds();
+
+            // If the signature is invalid, ignore it.
+            if (!(isValid_[index_] = _verifySignature(digest_, nodeId_, signatures_[index_].signature))) continue;
+
+            ++validSignatureCount_;
+        }
+
+        uint8 requiredSignatureCount_ = (INodeRegistryLike(nodeRegistry).canonicalNodesCount() / 2) + 1;
+
+        // Enforces that the number of valid signatures is greater than one more than half of the canonical node count.
+        if (validSignatureCount_ < requiredSignatureCount_) {
+            revert InsufficientSignatures(validSignatureCount_, requiredSignatureCount_);
+        }
+
+        validSigningNodeIds_ = new uint32[](validSignatureCount_);
+
+        uint256 writeIndex_ = 0;
+
+        for (uint256 index_; index_ < isValid_.length; ++index_) {
+            if (!isValid_[index_]) continue; // Skip invalid signatures.
+
+            unchecked {
+                validSigningNodeIds_[writeIndex_++] = signatures_[index_].nodeId;
+            }
+        }
+    }
+
+    /// @dev Returns true if the signature is from the owner of a canonical node.
+    function _verifySignature(
+        bytes32 digest_,
+        uint32 nodeId_,
+        bytes calldata signature_
+    ) internal view returns (bool isValid_) {
+        // If the node is not canonical, the signature is invalid.
+        if (!INodeRegistryLike(nodeRegistry).getIsCanonicalNode(nodeId_)) return false;
+
+        // Try to recover the signer from the signature. We don't want to revert if the signature is invalid.
+        address signer_ = ECDSA.tryRecoverCalldata(digest_, signature_);
+
+        // The signature is valid if the signer is not `address(0)` and the signer is the owner of the node.
+        return (signer_ != address(0)) && (signer_ == INodeRegistryLike(nodeRegistry).ownerOf(nodeId_));
```

Comment:
Something to think about here:

Everything we sign right now from the nodes (`OriginatorEnvelope`s) is signed with the node's `privateKey`, not necessarily with the private key that is the owner of the node NFT.

That allows node operators to hold their valuable NFT in one wallet, and put their private key on the node. If that private key is compromised, the attacker can send messages but can't steal money. 

This would change that. If we signed with and validated against the Node private key, we could avoid this problem and stop the private key from ever being exposed to the node.

Reply:
This would require the Node Registry to store the public key (or address) of each node's "signer".